File: go/src/levenshtein_test/levenshtein_test.go
package levenshtein_test
import (
	"testing"
	"levenshtein"
)

var levenshteinDistances = []struct {
	word1            string
	word2            string
	expectedDistance int
}{
	{"something", "something", 0},
	{"this", "has", 2},
	{"book", "back", 2},
	{"sitting", "kitten", 3},
	{"meilenstein", "levenshtein", 4},
	{"something", "", 9},
	{"", "something", 9},
}

func TestCalculateDistance(t *testing.T) {
	var techniques [2]levenshtein.LevenshteinDistanceTechnique
	techniques[0] = levenshtein.RecursiveLevenshteinDistanceTechnique{}
	techniques[1] = levenshtein.DynamicLevenshteinDistanceTechnique{}
	for _, technique := range techniques {
		for index, distanceSet := range levenshteinDistances {
			actualDistance := technique.CalculateDistance(distanceSet.word1, distanceSet.word2)
			if actualDistance != distanceSet.expectedDistance {
				t.Fatalf("at index %d, expected %d, got %d.", index, distanceSet.expectedDistance, actualDistance)
			}
		}
	}
}

File: ocaml/levenshtein_test.ml
open OUnit2

let test_same_distance_for_all_techniques test_ctxt =
  let distance_techniques = [|Levenshtein.recursive_distance; Levenshtein.dynamic_distance|]
  in
  for i = 0 to Array.length distance_techniques - 1 do
    assert_equal 0 (distance_techniques.(i) "something" "something");
    assert_equal 2 (distance_techniques.(i) "this" "has");
    assert_equal 2 (distance_techniques.(i) "book" "back");
    assert_equal 3 (distance_techniques.(i) "sitting" "kitten");
    assert_equal 4 (distance_techniques.(i) "meilenstein" "levenshtein");
    assert_equal 9 (distance_techniques.(i) "something" "");
    assert_equal 9 (distance_techniques.(i) "" "something")
  done

let suite =
"levenshtein_test">:::
 ["test_same_distance_for_all_techniques">:: test_same_distance_for_all_techniques]

let () = run_test_tt_main suite
